### ISSUE TITLE
BUG: Checking fp exception status causes false error

### ISSUE THREAD

**Issue by weiguozhi** [numpy/numpy#23381]
### Describe the issue:

After calling c function, numpy calls PyUFunc_checkfperr to check if there are any fp exceptions recorded and reports error if there are. 

In a default c environment, fp exceptions are disabled, when an fp exception condition is triggered, it will not cause a hardware exception, instead the exception is recorded in a status register, so PyUFunc_checkfperr can read out the status register and  reports error.

In this default environment, compiler can also do speculation optimization on fp instructions. For example in function hypergeometric_pdf_prime_loop_imp<long double>, there are following code
```
   ...
   if((p > 1) && (tools::max_value<T>() / p < result.value))
   ...
```
If the division "tools::max_value<T>() / p " is not speculated
```
 	callq	_ZN5boost4math6detail13integer_powerIeEET_RKS3_i@PLT
 	fldt	-124(%rbp)                      # 10-byte Folded Reload
 	fxch	%st(1)
 	fucomi	%st(1), %st
 	fstp	%st(1)
 	fldt	(%rbx)
	jbe	.LBB113_18
# %bb.17:                               #   in Loop: Header=BB113_12 Depth=1
 	fldt	-168(%rbp)                      # 10-byte Folded Reload
 	fdiv	%st(2), %st              // Not  speculated, executed after the first comparison
 	fxch	%st(1)
        ...
```
But a compiler can legally speculate the division before the first comparison and branch
```
 	callq	_ZN5boost4math6detail13integer_powerIeEET_RKS3_i@PLT
 	fldt	-124(%rbp)                      # 10-byte Folded Reload
 	fxch	%st(1)
 	fucomi	%st(1), %st
 	fstp	%st(1)
 	fldt	(%rbx)
 	fldt	-168(%rbp)                      # 10-byte Folded Reload
 	fdiv	%st(2), %st                  // Speculated! It is always executed no matter the first comparison result
	jbe	.LBB113_18
# %bb.17:                               #   in Loop: Header=BB113_12 Depth=1
 	fxch	%st(1)
```
It always computes the same result for any input data. But when user checks the fp status register, it may contain exception status generated by speculated instructions, and the speculated instruction result is not actually used by the intending function, so the exception status is not related to the computation, thus a false error is reported.

So here checking fp exception status has conflicts with compiler optimization.

### Reproduce the code example:

```python
NA
```

### Runtime information:

NA

**Comment by seberg** [comment]
clang defaults to not use `--trapping-math` which is supposed to make this predictable.  I am not sure what you are expecting?  The answer is "yes" that is exactly what happens.
Do you think NumPy is wrong for wanting to report floating point errors because a lot of C/C++ code doesn't bother (and clang defaults to not even trying)?

**Comment by weiguozhi** [comment]
The problem is with the default -fno-trapping-math, compiler may speculate fp instructions, these instructions results may not actually be used by the intending computing at run time, but they still impact sp status register.

So I think NumPy is wrong for reporting fp errors because there may be more fp instructions executed than minimum necessary because of optimization.

Let's study a simpler case
```
$clang++ -c -O2 t3.cpp --save-temps

void bar();

int foo(long double p, long double v, int *q) {
  *q = 3;
  if((p > 1.0) && (std::numeric_limits<long double>::max() / p < v))
    bar();

  return 0;
}
```

If we pass in a tiny value to p, the value is small enough so "max()/p" can generate a overflow exception condition. But a tiny p causes "(p > 1.0)" evaluates to false, so we don't expect the division will occur.

The division will occur only when we pass in a value larger than 1 to p, in this case, "max()/p" will be smaller than max, so we expect there will never be overflow exception status set by this function.

These analysis are based on non optimized compilation. For optimized compilation, compiler can freely move instructions when it can make sure the result is not impacted. The actual clang generated instructions are
```
# %bb.0:                                # %entry
    pushq   %rax
    .cfi_def_cfa_offset 16
    fldt    32(%rsp)
    fldt    16(%rsp)
    movl    $3, (%rdi)
    fld1
    fxch    %st(1)
    fucomi  %st(1), %st
    fstp    %st(1)
    fldt    .LCPI0_0(%rip)
    fdivp   %st, %st(1)             // speculated, always execute
    jbe .LBB0_3
# %bb.1:                                # %entry
    fxch    %st(1)
    fucompi %st(1), %st
    fstp    %st(0)
    fldz
    fldz
    jbe .LBB0_3
# %bb.2:                                # %if.then
    fstp    %st(1)
    fstp    %st(0)
    callq   _Z3barv@PLT
    fldz
    fldz
.LBB0_3:                                # %if.end
    fstp    %st(1)
    fstp    %st(0)
    xorl    %eax, %eax
    popq    %rcx
    .cfi_def_cfa_offset 8
    retq
```
You can see the div instruction in the entry block, so it is always executed. When you pass in a tiny value to p, the function's behavior is not impacted, except the fp exception status register now contains an overflow flag. If it is called from numpy, PyUFunc_checkfperr will report error.

**Comment by seberg** [comment]
@weiguozhi do you have a concrete issue?  We realize what compilers do, and they get it wrong more often than would be nice even with the right flags.  But we compile NumPy *with* trapping math and don't use the defaults.  And we have a test suite for many functions to get it right!

Some compilers will bug out, practically all have niche issues because there be dragons, this is all true, there are some ufuncs that will give a warning for inputs that should not warn, Mac tends to be worse than linux...

What exactly is the issue (or context of it) you have?  Are you writing a ufunc and unwilling to compile with flags that will ignore errors?  Do you think NumPy giving runtime  warnings for `NaN` or `inf` being created in ufuncs or casts is bad?

**Comment by weiguozhi** [comment]
> @weiguozhi do you have a concrete issue? We realize what compilers do, and they get it wrong more often than would be nice even with the right flags. But we compile NumPy _with_ trapping math and don't use the defaults. And we have a test suite for many functions to get it right!
> 

We encountered this problem in a large code base, and I failed to reduce it to a simple test case :(

Do you mean building NumPy with -ftrapping-math is the standard practice?

**Comment by seberg** [comment]
> Do you mean building NumPy with -ftrapping-math is the standard practice?

Clearly, our build system always sets it for a few years now.  There is an issue that e.g. due to `frompyfunc` NumPy ufuncs on `dtype=object` will call arbitrary Python code in a ufunc.  Arbitrary Python code of course has no reason to be clean about FPEs and NumPy happily reports them.

There is an issue about that one (but I just searched too long and didn't find it).  We could make sure to:
* Not check FPEs here (this is possible now), that isn't a 100% fix, but would hide them most of the time.
* We could actually restore FPE after calls into Python.  There are some corner cases where this may be relatively slow (The isuse had a solution for this, although I think I went too far with fenv, maybe just the flags is sufficient).

But maybe you are just compiling NumPy without those flags, in which case: yes, numpy needs it or the test suite will fail in many places with spurious warnings.
